core/src/main/scala/pureconfig/ConfigSource.scala

```diff
@@ -0,0 +1,320 @@
+   * `.withFallback` to specify optional config files to be merged (like `reference.conf`).
+   *
+   * @return a new `ConfigObjectSource` that provides the same config as this one, but falls back
+   *         to an empty config if it cannot be read.
+   */
+  def optional: ConfigObjectSource =
+    recoverWith { case ConfigReaderFailures(_: CannotRead, Nil) => Right(ConfigFactory.empty) }
+
+  /**
+   * Applies a function `f` if this source returns a failure, returning an alternative config in
+   * those cases.
+   *
+   * @param f the function to apply if this source returns a failure
+   * @return a new `ConfigObjectSource` that provides an alternative config in case this source
+   *         fails
+   */
+  def recoverWith(f: PartialFunction[ConfigReaderFailures, Result[Config]]): ConfigObjectSource =
+    new ConfigObjectSource(config.left.flatMap(f))
+}
+
+object ConfigObjectSource {
+
+  /**
+   * Creates a `ConfigObjectSource` from a `Result[Config]`. The provided argument is allowed
+   * to change value over time.
+   *
+   * @param conf the config to be provided by this source
+   * @return a `ConfigObjectSource` providing the given config.
+   */
+  def apply(conf: => Result[Config]): ConfigObjectSource =
+    new ConfigObjectSource(conf)
+
+  /**
+   * Creates a `ConfigObjectSource` from a `ConfigObjectCursor`.
+   *
+   * @param cur the cursor to be provided by this source
+   * @return a `ConfigObjectSource` providing the given cursor.
+   */
+  def fromCursor(cur: ConfigObjectCursor): ConfigObjectSource =
+    new ConfigObjectSource(Right(cur.value.toConfig))
+
+  /**
+   * Creates a `ConfigObjectSource` from a `FluentConfigCursor`.
+   *
+   * @param cur the cursor to be provided by this source
+   * @return a `ConfigObjectSource` providing the given cursor.
+   */
+  def fromCursor(cur: FluentConfigCursor): ConfigObjectSource =
+    new ConfigObjectSource(cur.asObjectCursor.right.map(_.value.toConfig))
+}
+
+/**
+ * Object containing factory methods for building `ConfigSource`s.
+ *
+ * The sources provided here use Typesafe Config configs created from files, resources, URLs or
+ * strings. It also provides sources that delegate the loading component to Typesafe Config, to
+ * leverage reference configs and overrides, making it easy to switch from using `ConfigFactory`
+ * to `ConfigSource`.
+ *
+ * Other PureConfig modules may provide other ways or building config sources (e.g. for different
+ * config formats or data sources).
+ */
+object ConfigSource {
+
+  /**
+   * A config source for the default loading process in Typesafe Config. Typesafe Config stacks
+   * `reference.conf` resources provided by libraries, application configs (by default
+   * `application.conf` in resources) and system property overrides, resolves them and merges them
+   * into a single config. This source is equivalent to
+   * `defaultOverrides.withFallback(defaultApplication).withDefault(defaultReference)`.
```

Comment:
This should be `withFallback`:

```suggestion
   * `defaultOverrides.withFallback(defaultApplication).withFallback(defaultReference)`.
```